### PR TITLE
Make LS Config-API extensible for extension devs

### PR DESCRIPTION
**Pull request by rasika**
Fixes https://github.com/ballerina-platform/ballerina-lang/issues/26318

**1. First, you need to write you own config class extending LSClientConfig.**
```java 
public class YourExtendedConfig extends LSClientConfig {
    private final DataMapperConfig dataMapper;

    protected YourExtendedConfig() {
        super();
        this.dataMapper = new DataMapperConfig();
    }

    public DataMapperConfig getDataMapper() {
        return dataMapper;
    }
}
```

**2. Now you can read the data mapper config as below;**

```java
LSClientConfigHolder.getInstance().getConfigAs(YourExtendedConfig.class).getDataMapper();
```
**Or you can register a callback listener;**
```java
LSClientConfigHolder.getInstance().register(new ClientConfigListener() {
        @Override
        public void didChangeJson(JsonElement oldConfigJson, JsonElement newConfigJson) {
                
        }
    }
);
```

## Check List 
- [x] Read the [Contributing Guide](https://github.com/ballerina-platform/ballerina-lang/blob/master/CONTRIBUTING.md)
- [ ] Updated Change Log
- [ ] Checked Tooling Support (#<Issue Number>)
- [ ] Added necessary tests
  - [ ] Unit Tests
  - [ ] Spec Conformance Tests
  - [ ] Integration Tests
  - [ ] Ballerina By Example Tests
- [ ] Increased Test Coverage   
- [ ] Added necessary documentation  
  - [ ] API documentation 
  - [ ] Module documentation in Module.md files
  - [ ] Ballerina By Examples

### CODE REVIEW

File: language-server/modules/langserver-compiler/src/main/java/org/ballerinalang/langserver/compiler/config/ExtendedClientConfigListener.java [hunk 1]
```diff
@@ -0,0 +1,44 @@
...
+
+import com.google.gson.JsonElement;
+
+/**
+ * Represents a ballerina client config change listener.
+ *
+ * @since 2.0.0
+ */
+@FunctionalInterface
+public interface ExtendedClientConfigListener extends BaseClientConfigListener {
```
Comment (nadeeshaan) [inline language-server/modules/langserver-compiler/src/main/java/org/ballerinalang/langserver/compiler/config/ExtendedClientConfigListener.java:26]: Can't we depend only on the `didChangeConfig`? If so we do not need to have two interfaces for `ExtendedClientConfigListener` and `ClientConfigListener`. Now the hierarchy have three interfaces.

Comment (rasika) [reply language-server/modules/langserver-compiler/src/main/java/org/ballerinalang/langserver/compiler/config/ExtendedClientConfigListener.java]: We do need both `didChangeConfig` and `didChangeJson` for different use cases. As discussed during the call, I'll remove the inherited interfaces for simplicity.

File: misc/ballerinalang-data-mapper/src/main/java/org/ballerinalang/datamapper/config/LSClientExtendedConfig.java [hunk 2]
```diff
@@ -0,0 +1,41 @@
...
+package org.ballerinalang.datamapper.config;
+
+import org.ballerinalang.langserver.compiler.config.LSClientConfig;
+
+/**
+ * Extended Ballerina Client Configuration.
+ *
+ * @since 2.0.0
+ */
+public class LSClientExtendedConfig extends LSClientConfig {
```
Comment (nadeeshaan) [inline misc/ballerinalang-data-mapper/src/main/java/org/ballerinalang/datamapper/config/LSClientExtendedConfig.java:25]: Shall we rename this to `DatamapperClientConfig` or similar one, without using the extended word?

**Review by nadeeshaan — APPROVED** [review]

**Review by Ayodhya94 — APPROVED** [review]
Tested code action locally and it is working as expected

File: misc/ballerinalang-data-mapper/src/main/java/org/ballerinalang/datamapper/config/LSClientExtendedConfig.java (hunk 2 above)
Comment (rasika) [reply misc/ballerinalang-data-mapper/src/main/java/org/ballerinalang/datamapper/config/LSClientExtendedConfig.java:25]: Let's fix this in a separate PR.

**Pull request merged**